### ISSUE TITLE
Linker error when using folly (undefined reference to ScopeGuardImplBase::warnAboutToCrash())

### ISSUE THREAD

**Issue by danrthompson** [facebook/folly#751]
When I try to build our codebase with folly I'm getting a linker error. I've narrowed it down to where it fails with this program:

```
#include "folly/dynamic.h"
#include "folly/json.h"

int main(int argc, char** argv) {
  folly::dynamic dyn;
  folly::toJson(dyn);
  return 0;
}
```

When I try to statically link the folly library to this program I get this error:
```
ext/folly/upstream/folly/ScopeGuard.h:45: error: undefined reference to 'folly::detail::ScopeGuardImplBase::warnAboutToCrash()'
```

Folly builds fine, and if I dynamically link it to the program it works. Our folly repo is at commit `e869664`

I was able to fix the error by moving the ScopeGuard.cpp implementation into ScopeGuard.h. ScopeGuard.cpp only contains an implementation for one method, `ScopeGuardImplBase::warnAboutToCrash()`, so basically my change was to just inline the implementation of that method in ScopeGuard.h. I thought about creating a pull request, but I'm not 100% clear on why this is happening or why my fix solves it and I wanted to bring up the issue first. I think it might have to do with the fact that `warnAboutToCrash()` is called by `runAndWarnAboutToCrashOnException()`, which is a templated function, but I'm not sure.

Any insight would be much appreciated.

**Comment by yfeldblum** [comment]
Are you including the folly header at the same rev as the linked-to folly library was built from?

**Comment by danrthompson** [comment]
Sorry I don't understand your question. Our folly library is at commit e869664. It builds fine. And my little sample program compiles fine. But then when I statically link the folly library into the sample program it gives that error. Same with our actual codebase. I think it might be due to an error in the code in ScopeGuard.h - and when I made the change I described to ScopeGuard.h it solves the problem - but I'm not sure
